pkg/virt-controller/services/template.go

```diff
@@ -1494,168 +958,714 @@ func (t *templateService) RenderHotplugAttachmentPodTemplate(volumes []*v1.Volum
+								MatchLabels: ownerPod.GetLabels(),
+							},
+							TopologyKey: "kubernetes.io/hostname",
+						},
+					},
+				},
+			},
+			Volumes: []k8sv1.Volume{
+				{
+					Name: volume.Name,
+					VolumeSource: k8sv1.VolumeSource{
+						PersistentVolumeClaim: &k8sv1.PersistentVolumeClaimVolumeSource{
+							ClaimName: pvcName,
+							ReadOnly:  false,
+						},
+					},
+				},
+				{
+					Name: "hotplug-disks",
+					VolumeSource: k8sv1.VolumeSource{
+						EmptyDir: &k8sv1.EmptyDirVolumeSource{},
+					},
+				},
+			},
+			HostNetwork:                   true,
+			TerminationGracePeriodSeconds: &zero,
+		},
+	}
+
+	if isBlock {
+		pod.Spec.Containers[0].VolumeDevices = []k8sv1.VolumeDevice{
+			{
+				Name:       volume.Name,
+				DevicePath: "/dev/hotplugblockdevice",
+			},
+		}
+		pod.Spec.SecurityContext = &k8sv1.PodSecurityContext{
+			RunAsUser: &[]int64{0}[0],
+		}
+	} else {
+		pod.Spec.Containers[0].VolumeMounts = append(pod.Spec.Containers[0].VolumeMounts, k8sv1.VolumeMount{
+			Name:      volume.Name,
+			MountPath: "/pvc",
+		})
+	}
+	return pod, nil
+}
+
+func renderLaunchManifestInitDefaultVolumes(t *templateService, vmi *v1.VirtualMachineInstance) (v []k8sv1.Volume, vd []k8sv1.VolumeDevice, vmo []k8sv1.VolumeMount, hotplugVolumes map[string]bool) {
+
+	addVolume := func(name string) {
+		v = append(v, k8sv1.Volume{
+			Name: name,
+			VolumeSource: k8sv1.VolumeSource{
+				EmptyDir: &k8sv1.EmptyDirVolumeSource{},
+			},
+		})
+	}
+	addVolumeMount := func(name, path string, mode *k8sv1.MountPropagationMode) {
+		vmo = append(vmo, k8sv1.VolumeMount{
+			Name:             name,
+			MountPath:        path,
+			MountPropagation: mode,
+		})
+	}
+
+	hotplugVolumes = make(map[string]bool)
+	for _, volumeStatus := range vmi.Status.VolumeStatus {
+		if volumeStatus.HotplugVolume != nil {
+			hotplugVolumes[volumeStatus.Name] = true
+		}
+	}
+
+	addVolumeMount("private", util.VirtPrivateDir, nil)
+	addVolume("private")
+
+	addVolumeMount("public", util.VirtShareDir, nil)
+	addVolume("public")
+
+	addVolumeMount("ephemeral-disks", t.ephemeralDiskDir, nil)
+
+	prop := k8sv1.MountPropagationHostToContainer
+
+	addVolumeMount("container-disks", t.containerDiskDir, &prop)
+	if !vmi.Spec.Domain.Devices.DisableHotplug {
+		addVolumeMount("hotplug-disks", t.hotplugDiskDir, &prop)
+	}
+
+	addVolumeMount("libvirt-runtime", "/var/run/libvirt", nil)
+
+	// virt-launcher cmd socket dir
+	addVolumeMount("sockets", filepath.Join(t.virtShareDir, "sockets"), nil)
+	addVolume("sockets")
+
+	return
+}
+
+func renderLaunchManifestVolumes(t *templateService, vmi *v1.VirtualMachineInstance, volumes []k8sv1.Volume, volumeDevices []k8sv1.VolumeDevice, volumeMounts []k8sv1.VolumeMount, hotplugVolumes map[string]bool, namespace string) ([]k8sv1.Volume, []k8sv1.VolumeDevice, []k8sv1.VolumeMount, []k8sv1.LocalObjectReference, string, error) {
+	var imagePullSecrets []k8sv1.LocalObjectReference
+	serviceAccountName := ""
```

Comment:
is it okay to have this defined twice? can it be changed anywhere? if it gets updated one place but not the other will it break something?

Reply:
TBH I didn't understand what you're referring to. What is defined twice?